lib/activerecord_autoreplica.rb

```diff
@@ -69,32 +88,73 @@ def self.using_read_replica_pool(replica_connection_pool)
     in_replica_context(replica_connection_pool){ yield }
   end
 
-  def self.in_replica_context(handler_params, handler_class=ConnectionHandler)
-    return yield if Thread.current[:autoreplica] # This method should not be reentrant
+  def self.in_replica_context(read_pool)
+    return yield if current_read_pool # This method should not be reentrant
+
+    # There is a pontential race condition here in a threaded environment, but
+    # in the worst case the handler will be set up twice. This shouldn't affect
+    # operations.
```

Comment:
Why not have a mutex then? It's cheap, and is what the original code tries to do (hopefully, will make your diff likelier to get accepted upstream)